### ISSUE TITLE
no-extra-parens and no-cond-assign ("except-parens") don't play well

### ISSUE THREAD

**Issue by spadgos** [eslint/eslint#3317]
Maybe I'm missing something, but it seems to be impossible to use the `no-extra-parens` rule alongside `no-cond-assign`, with its exception for conditionals wrapped in parentheses (the default).

That is, there's no way _(that I've found)_ to express this code without an error with both of the rules turned on:

```js
while ((foo = bar())) { ... }
```

`no-cond-assign` by default requires assignments within conditionals to be wrapped, and errors if it isn't. `no-extra-parens` errors if you _do_ use the parens.

Just as it has exceptions for regexes, I would suggest that `no-extra-parens` should not complain about using parens to wrap an assignment in a conditional.

---

Info for @eslintbot: Using eslint 1.0.0, with the code above, and `no-cond-assign` and `no-extra-parens` turned on.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by michaelficarra** [comment]
I would not want this to be the default, but that sounds like a useful option.

**Comment by nzakas** [comment]
That's an interesting conundrum. We probably need an option on no-extra-parens to allow that.

**Comment by alberto** [comment]
Shouldn't be this the default behaviour? Adding the extra parens seems explicit enough to me. Anyway, if we were to add it, should it be a third option (for example `all-except-cond-assign`) or an additional argument for `all`?

**Comment by nzakas** [comment]
Are you saying we should always allow parens when there's an assignment?

**Comment by alberto** [comment]
Yes, allow double parens around the assignment inside conditionals, so these would be valid.
```
"while ((foo = bar())) {}",
"if ((foo = bar())) {}",
"do; while ((foo = bar()))",
```
 Do you think that would be a problem?

**Comment by michaelficarra** [comment]
This needs to be behind an option though.

**Comment by alberto** [comment]
@michaelficarra can you elaborate on why? Maybe I'm missing something

**Comment by nzakas** [comment]
The only problem is that those are unnecessary parens, so we are kind of breaking the rule if we allow it by default. Adding as an option would let people opt-in.

**Comment by michaelficarra** [comment]
Exactly.

**Comment by alberto** [comment]
I proposed that because that's what the rule already does with Regexps and IIFEs.

Is the name `all-except-cond-assign` ok for you?

**Comment by michaelficarra** [comment]
Yeah, I really wish those were behind flags as well, but it's probably too late for that now.
